src/pg_copy.c

```diff
@@ -0,0 +1,696 @@
+		}
+		else
+		{
+			failedPlacementList = lappend(failedPlacementList, taskPlacement);
+			ereport(WARNING, (errcode(ERRCODE_IO_ERROR),
+							  errmsg("Failed to connect to node %s:%d", 
+									 nodeName, taskPlacement->nodePort)));
+		}
+	}
+
+	/* if all placements failed, error out */
+	if (list_length(failedPlacementList) == list_length(finalizedPlacementList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_IO_ERROR),
+						errmsg("Could not copy to any active placements for shard %ld",
+							   (long) shardId)));
+
+		/* otherwise, mark failed placements as inactive: they're stale */
+	}
+	foreach(failedPlacementCell, failedPlacementList)
+	{
+		ShardPlacement *failedPlacement = (ShardPlacement *) lfirst(failedPlacementCell);
+
+		UpdateShardPlacementRowState(failedPlacement->id, STATE_INACTIVE);
+	}
+	shardConnections->replicaCount = placementCount;
+}
+
+
+static List *
+HTABToList(HTAB *hash)
+{
+	HASH_SEQ_STATUS hashCursor;
+	void *entry;
+	List *list = NULL;
+
+	hash_seq_init(&hashCursor, hash);
+	while ((entry = hash_seq_search(&hashCursor)) != NULL)
+	{
+		list = lappend(list, entry);
+	}
+	return list;
+}
+
+
+/*
+ * Handle copy to/from distributed table
+ */
+void
+PgShardCopy(CopyStmt *copyStatement, char const *query)
+{
+	RangeVar *relation = copyStatement->relation;
+	ListCell *shardIntervalCell = NULL;
+	List *shardIntervalList = NULL;
+	char *relationName = NULL;
+	PgShardTransactionManager const *transactionManager =
+		&PgShardTransManagerImpl[PgShardCurrTransManager];
+	bool failOK = true;
+	Oid tableId = RangeVarGetRelid(relation, NoLock, failOK);
+	HTAB *shardToConn = NULL;
+	MemoryContext tupleContext = NULL;
+	CopyState copyState = NULL;
+	bool nextRowFound = true;
+	TupleDesc tupleDescriptor = NULL;
+	uint32 columnCount = 0;
+	Datum *columnValues = NULL;
+	bool *columnNulls = NULL;
+	Var *partitionColumn = NULL;
+	Oid columnOid = 0;
+	OpExpr *equalityExpr = NULL;
+	List *whereClauseList = NULL;
+	List *prunedList = NULL;
+	Node *rightOp = NULL;
+	Const *rightConst = NULL;
+	ShardId failedShard = INVALID_SHARD_ID;
+	Relation rel = NULL;
+	StringInfo lineBuf;
+	ShardConnections *shardConnections = NULL;
+	List *shardConnectionsList = NULL;
+	List **shardListCache = NULL;
+	Datum partitionColumnValue = 0;
+	int i = 0;
+	TypeCacheEntry *typeEntry = NULL;
+	FmgrInfo *hashFunction = NULL;
+	int hashedValue = 0;
+	int shardCount = 0;
+	int shardHashCode = 0;
+	uint32 hashTokenIncrement = 0;
+
+	relationName = get_rel_name(tableId);
+
+	shardIntervalList = LookupShardIntervalList(tableId);
+	if (shardIntervalList == NIL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("could not find any shards for query"),
+						errdetail("No shards exist for distributed table \"%s\".",
+								  relationName),
+						errhint("Run master_create_worker_shards to create shards "
+								"and try again.")));
+	}
+	if (!copyStatement->is_from) /* Construct query selecting data from this relation */
```

Comment:
I have a slight preference for moving this code block outside of this function to separate concerns. What if we introduce an intermediate function that distinguishes between COPY FROM and COPY TO?

    void
    PgShardCopy(CopyStmt *copyStatement, char const *query)
    {
        if (copyStatement->is_from)
        {
            PgShardCopyFrom(copyStatement, query);
        }
        else
        {
            PgShardCopyTo(copyStatement, query);
        }
    }

PgShardCopyTo would contain the code below.